jax/lax/lax.py

```diff
@@ -1401,8 +1401,18 @@ def full(shape: Shape, fill_value: Array, dtype: Optional[DType] = None) -> Arra
     fill_value = xla.device_put_p.bind(convert_element_type(fill_value, dtype))
   return broadcast(fill_value, shape)
 
+# TODO(mattjj) remove, we only need this function because _device_put_raw is
```

Comment:
IIRC, the reason for that choice is efficiency. It turns out that avoiding the generic primitive dispatch code is faster and it matters for putting things on device. We should probably fix that.

Reply:
@hawkinsp Agree. This would allow much cleaner implementations of new backends in general. We could open an issue after merging.